iroh-net/src/hp/portmapper.rs

```diff
@@ -14,41 +21,151 @@ pub struct ProbeResult {
+            upnp,
+        })
+    }
+
+    /// Checks if the last seen upnp gateway should still be trusted as available.
+    ///
+    /// Returns false otherwise, or if there is no cached gateway.
+    fn upnp_available_from_cache(&self) -> bool {
+        self.last_upnp_gateway_addr
+            .as_ref()
+            .map(|(_gateway_addr, last_probed)| {
+                *last_probed + AVAILABILITY_TRUST_DURATION > Instant::now()
+            })
+            .unwrap_or_default()
+    }
+
+    /// Probes for UPnP gateways.
+    async fn probe_upnp_available(&mut self) -> bool {
+        match upnp::probe_available().await {
+            Ok(gateway_addr) => {
+                let old_gateway = self
+                    .last_upnp_gateway_addr
+                    .replace((gateway_addr, Instant::now()));
+                if let Some((old_gateway_addr, _last_seen)) = old_gateway {
+                    if old_gateway_addr != gateway_addr {
+                        tracing::debug!(
+                            "upnp gateway changed from {old_gateway_addr} to {gateway_addr}"
+                        );
+                        // TODO(@divagant-martian): tailscale does not invalidate the mappings here. Why?
+                    }
+                }
+                true
+            }
+            Err(e) => {
+                info!("upnp probe failed {e}");
+                // invalidate last seen gateway and time
+                self.last_upnp_gateway_addr = None;
+                false
+            }
+        }
     }
 
     /// Updates the local port number to which we want to port map UDP traffic.
-    pub async fn set_local_port(&self, _local_port: u16) {
-        // TODO:
+    /// Invalidates the current mapping if any.
+    pub async fn set_local_port(&mut self, local_port: u16) {
+        // TODO(@divagant-martian): this should never be 0
+        if self.local_port != local_port {
+            self.local_port = local_port;
+            let _ = self.invalidate_mapping().await;
+        }
     }
 
     /// Quickly returns with our current cached portmapping, if any.
     /// If there's not one, it starts up a background goroutine to create one.
     /// If the background goroutine ends up creating one, the `on_change` hook registered with the
     /// `Client::new` constructor (if any) will fire.
-    pub async fn get_cached_mapping_or_start_creating_one(&self) -> Option<SocketAddr> {
-        // TODO:
-        None
+    // TODO(@divagant-martian): fix docs. Also, this will probably only ever return ipv4 addresses.
+    pub async fn get_cached_mapping_or_start_creating_one(&mut self) -> Option<SocketAddr> {
+        // TODO(@divagant-martian): this has a lock ensure just one mapping attempt is underway
+        if let Some(mapping) = &mut self.current_mapping {
+            let now = Instant::now();
+            if now <= mapping.good_until() {
+                // TODO(@divagant-martian): this would go in a goroutine
```

Comment:
Given this structure, it would probably be nicest to avoid locks etc, to just have a small actor like setup, for the mapping. Even though it is a bit more boilerplate, being able to avoid locks usually pays of in this kind of code in my experience.

Reply:
Agree. In general the way portmapping is wired can get more love but it's understandable this was not clear until just now